// puppeteer-screen-recorder/build/main/spec/PuppeteerScreenRecorder.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = require("path");
const ava_1 = __importDefault(require("ava"));
const puppeteer_1 = __importDefault(require("puppeteer"));
const __1 = require("../");
const launchBrowser = async () => {
    const puppeteerPath = process.env['PUPPETEER_EXECUTABLE_PATH'];
    if (!puppeteerPath) {
        throw new Error('Please configure puppeteer path: export PUPPETEER_EXECUTABLE_PATH=<path>');
    }
    const browser = await puppeteer_1.default.launch(Object.assign({ headless: true }, (puppeteerPath ? { executablePath: puppeteerPath } : {})));
    return browser;
};
ava_1.default('case 1a --> Happy Path: Should be able to create a new screen-recording session', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const outputVideoPath = './test-output/test/video-recorder/testCase1.mp4';
    const recorder = new __1.PuppeteerScreenRecorder(page);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    await page.goto('https://google.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 1a --> Happy Path: Should be able to create a new screen-recording session using mp4 format', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const outputVideoPath = './test-output/test/video-recorder/testCase1.mp4';
    const recorder = new __1.PuppeteerScreenRecorder(page);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    await page.goto('https://google.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 1b --> Happy Path: Should be able to create a new screen-recording session using mov', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const outputVideoPath = './test-output/test/video-recorder/testCase1.mov';
    const recorder = new __1.PuppeteerScreenRecorder(page);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    await page.goto('https://google.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 1c --> Happy Path: Should be able to create a new screen-recording session using webm', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const outputVideoPath = './test-output/test/video-recorder/testCase1.webm';
    const recorder = new __1.PuppeteerScreenRecorder(page);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    await page.goto('https://google.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 1d --> Happy Path: should be to get the total duration of recording using avi', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const outputVideoPath = './test-output/test/video-recorder/testCase2.avi';
    const recorder = new __1.PuppeteerScreenRecorder(page);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    await page.goto('https://google.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    const duration = recorder.getRecordDuration();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    console.log('duration', duration);
    assert.is(duration !== '00:00:00:00', true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 2 --> Happy Path: testing video recording with video frame width, height and aspect ratio', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const options = {
        followNewTab: false,
        videoFrame: {
            width: 1024,
            height: 1024,
        },
        aspectRatio: '4:3',
    };
    const outputVideoPath = './test-output/test/video-recorder/testCase4.mp4';
    const recorder = new __1.PuppeteerScreenRecorder(page, options);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('test 3 -> Error Path: should throw error if an invalid savePath argument is passed for start method', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    try {
        const outputVideoPath = './test-output/test/video-recorder/';
        const recorder = new __1.PuppeteerScreenRecorder(page);
        await recorder.start(outputVideoPath);
        /** execute */
        await page.goto('https://github.com', { waitUntil: 'load' });
        await page.goto('https://google.com', { waitUntil: 'load' });
        /** clear */
        await recorder.stop();
    }
    catch (error) {
        assert.true(error.message === 'File format is not supported');
    }
});
ava_1.default('case 4 --> Happy Path: Should be able to create a new screen-recording session using streams', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const outputVideoPath = './test-output/test/video-recorder/testCase4.mp4';
    try {
        fs_1.default.mkdirSync(path_1.dirname(outputVideoPath), { recursive: true });
    }
    catch (e) {
        console.error(e);
    }
    const fileWriteStream = fs_1.default.createWriteStream(outputVideoPath);
    const recorder = new __1.PuppeteerScreenRecorder(page);
    await recorder.startStream(fileWriteStream);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    await page.goto('https://google.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
    fileWriteStream.on('end', () => {
        assert.is(fileWriteStream.writableFinished, true);
    });
});
ava_1.default('case 5a --> Happy Path: testing video recording with video frame width, height and autopad color', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const options = {
        followNewTab: false,
        videoFrame: {
            width: 1024,
            height: 1024,
        },
        autopad: {
            color: 'gray',
        },
    };
    const outputVideoPath = './test-output/test/video-recorder/testCase5a.mp4';
    const recorder = new __1.PuppeteerScreenRecorder(page, options);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 5b --> Happy Path: testing video recording with video frame width, height and autopad color as hex code', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const options = {
        followNewTab: false,
        videoFrame: {
            width: 1024,
            height: 1024,
        },
        autopad: {
            color: '#008000',
        },
    };
    const outputVideoPath = './test-output/test/video-recorder/testCase5b.mp4';
    const recorder = new __1.PuppeteerScreenRecorder(page, options);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 5c --> Happy Path: testing video recording with video frame width, height and default autopad color', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const options = {
        followNewTab: false,
        videoFrame: {
            width: 1024,
            height: 1024,
        },
        autopad: {},
    };
    const outputVideoPath = './test-output/test/video-recorder/testCase5c.mp4';
    const recorder = new __1.PuppeteerScreenRecorder(page, options);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
ava_1.default('case 6 --> Happy Path: should create a video with a custom crf', async (assert) => {
    /** setup */
    const browser = await launchBrowser();
    const page = await browser.newPage();
    const options = {
        followNewTab: false,
        videoFrame: {
            width: 1024,
            height: 1024,
        },
        videoCrf: 0,
    };
    const outputVideoPath = './test-output/test/video-recorder/testCase6.mp4';
    const recorder = new __1.PuppeteerScreenRecorder(page, options);
    const recorderValue = await recorder.start(outputVideoPath);
    /** execute */
    await page.goto('https://github.com', { waitUntil: 'load' });
    /** clear */
    const status = await recorder.stop();
    await browser.close();
    /** assert */
    assert.is(recorderValue instanceof __1.PuppeteerScreenRecorder, true);
    assert.is(status, true);
    assert.is(fs_1.default.existsSync(outputVideoPath), true);
});
